core/server/proxy/src/test/java/alluxio/StreamCacheTest.java

```diff
@@ -47,70 +48,127 @@ public void concurrentOperations() throws Exception {
+
+    // Concurrent get, put, and invalidate.
+    class GetPutInvalidate implements Runnable {
+      private int mThreadIndex;
+
+      GetPutInvalidate(int threadIndex) {
+        mThreadIndex = threadIndex;
+      }
+
       @Override
       public void run() {
-        Assert.assertSame(os, streamCache.getOutStream(osId.get()));
+        Assert.assertSame(is, streamCache.getInStream(isIDs.get(mThreadIndex)));
+        Assert.assertSame(os, streamCache.getOutStream(osIDs.get(mThreadIndex)));
+        int isId = streamCache.put(is);
+        int osId = streamCache.put(os);
+        Assert.assertSame(is, streamCache.invalidate(isIDs.get(mThreadIndex)));
+        Assert.assertSame(os, streamCache.invalidate(osIDs.get(mThreadIndex)));
+        Assert.assertSame(is, streamCache.invalidate(isId));
+        Assert.assertSame(os, streamCache.invalidate(osId));
       }
-    };
-    Thread getIsThread = new Thread(getIs);
-    Thread getOsThread = new Thread(getOs);
-    getIsThread.start();
-    getOsThread.start();
-    getIsThread.join();
-    getOsThread.join();
+    }
+    for (int i = 0; i < numThreads; i++) {
+      threads[i] = new Thread(new GetPutInvalidate(i));
+    }
+    for (Thread thread : threads) {
+      thread.start();
+    }
+    for (Thread thread : threads) {
+      thread.join();
+    }
+    for (int i = 0; i < numThreads; i++) {
+      Assert.assertNull(streamCache.getInStream(isIDs.get(i)));
+      Assert.assertNull(streamCache.getOutStream(osIDs.get(i)));
+    }
+    Assert.assertEquals(0, streamCache.size());
 
-    // Concurrent get, put, and invalidate.
-    final Integer oldIsId = isId.get();
-    Thread invalidateIsThread = new Thread(new Runnable() {
+    // Concurrent put, get, and invalidate.
+    Runnable putGetInvalidate = new Runnable() {
       @Override
       public void run() {
-        Assert.assertSame(is, streamCache.invalidate(oldIsId));
+        int isId = streamCache.put(is);
+        int osId = streamCache.put(os);
+        Assert.assertSame(is, streamCache.getInStream(isId));
+        Assert.assertSame(os, streamCache.getOutStream(osId));
+        Assert.assertSame(is, streamCache.invalidate(isId));
+        Assert.assertSame(os, streamCache.invalidate(osId));
       }
-    });
-    invalidateIsThread.start();
-    putIsThread = new Thread(putIs);
-    putIsThread.start();
-    getOsThread = new Thread(getOs);
-    getOsThread.start();
-    invalidateIsThread.join();
-    putIsThread.join();
-    getOsThread.join();
-    Assert.assertNull(streamCache.getInStream(oldIsId));
-    Assert.assertSame(is, streamCache.getInStream(isId.get()));
-    Assert.assertSame(os, streamCache.getOutStream(osId.get()));
+    };
+    for (int i = 0; i < numThreads; i++) {
+      threads[i] = new Thread(putGetInvalidate);
+    }
+    for (Thread thread : threads) {
+      thread.start();
+    }
+    for (Thread thread : threads) {
+      thread.join();
+    }
+    Assert.assertEquals(0, streamCache.size());
   }
```

Comment:
is there any value in having the separate put and get tests if the last test ends up doing put, get, invalidate? if not, maybe you can just have that one and run it with higher concurrency ... (choose concurrency resulting in 5-10 second runtime)